src/proc.cpp

```diff
@@ -568,22 +567,43 @@ static bool process_clean_after_marking(bool allow_interactive) {
             }
             proc_fire_event(L"JOB_EXIT", event_type_t::job_exit, j->job_id, 0);
 
-            itr = jobs().erase(itr);
-            erased = true;
+            erase_list.push_back(j.get());
         } else if (j->is_stopped() && !j->get_flag(job_flag_t::NOTIFIED)) {
             // Notify the user about newly stopped jobs.
             if (!j->get_flag(job_flag_t::SKIP_NOTIFICATION)) {
-                print_job_status(j, JOB_STOPPED);
+                print_job_status(j.get(), JOB_STOPPED);
                 found = true;
             }
             j->set_flag(job_flag_t::NOTIFIED, true);
         }
     }
 
-    if (found) fflush(stdout);
+    if (!erase_list.empty()) {
+        // The intersection of the two lists is typically O(m*n), but we know the order
+        // of the entries in erase_list is the same as their matches in jobs(), so we can
+        // use that to our advantage.
+        auto to_erase = erase_list.begin();
+        jobs().erase(std::remove_if(jobs().begin(), jobs().end(),
+                    [&to_erase](const shared_ptr<job_t> &j) {
+            if (to_erase == erase_list.end()) {
+                return false;
+            }
+            if (*to_erase == j.get()) {
+                ++to_erase;
+                return true;
+            }
+            return false;
+        }), jobs().end());
+        assert(to_erase == erase_list.end()
```

Comment:
My feeling is that this assertion is invalid. I think the following might happen:

1. Some jobs are reapable.
2. During reaping, we post JOB_EXIT. This triggers a user function which re-entrantly reaps jobs again.
3. Oops the job we planned to reap has already been reaped, we trip the assertion.

We should have better test coverage of JOB_EXIT, but for now I think this assertion is wrong.

Reply:
Hmm, that's possible. It's mainly there in case of an (extremely hypothetical) `std::remove_if` implementation that does not iterate in-order, but as it's not the end of the world even in such a highly unlikely scenario, it's OK to drop the assertion.

Reply:
At the end of the day, the job list is a shared global object and if we want to support reentrance for functions that involve its manipulation (or the manipulation of properties of its contents) we're going to need add a lock there. It's a high-read/low-write probability situation, it might be worth using an RW lock whenever we do go down this path.

Reply:
At present, there's a poor-mans lock around the entire function (which should at the very least be converted to an interlocked CAS) preventing any of those scenarios from actually happening, which is why I felt OK making the erase list static (for no other reason than the performance benefits of not allocating on the heap each time). 

I updated the code to only post the `JOB_EXIT` once the job has actually been erased, which should prevent/forestall any such scenarios from arising in the first place, but as the job list represents globally shared state, any manipulations to or enumerations of it fundamentally needs to be protected by *some* form locking, unless all access (reads and writes, both) is guaranteed to come from the main thread (and surround with `ASSERT_IS_MAIN_THREAD()` checks). It's a high-read/low-write probability structure, so it would probably be worth going with an RW lock here rather than a standard mutex if/when we ever implement such locking.

The alternative would be some form of lock-free access where we guarantee that a) enumeration is always safe at all times, even in the middle of an addition/removal (so stricter safety semantics than what `std::list` provides as pointer writes aren't guaranteed atomic on all the platforms we support), but more importantly, that the side effects of accessing possibly outdated data are well understood and built into the logic, which is going to be incredibly hard as fish was simply not designed to do so.